File: 3/3.js
// // The prime factors of 13195 are 5, 7, 13 and 29.

// What is the largest prime factor of the number 600851475143 ?

function findLargestPrime(integer){
  var factorArray = [];
  for(var i = 2; i < integer; i++){
    if(integer % i === 0){
      factorArray.push(i);
    }
  }
  largestFactor = factorArray[factorArray.length - 1];
  for(var x = 0; x < factorArray.length; x++){
    for(var i = 2; i < largestFactor + 1; i++){
      if(i != factorArray[x] && factorArray[x] % i === 0){
        factorArray.splice(x, 1)[0];
        x -= 1;
        i = largestFactor;
      }
    }
  }
  return factorArray[factorArray.length - 1]
}

findLargestPrime(600851475143)
// => 6857

File: 3/3.rb
# The prime factors of 13195 are 5, 7, 13 and 29

# What is the largest prime factor of the number 600851475143 ?

def find_largest_prime(integer)
  factor_array = []
  potential_factor_array = (2..integer).to_a
  potential_factor_array.each do |factor|
    if integer % factor == 0
      factor_array << factor
    end
  end
  largest_factor = factor_array.last
  potential_factor_array = (2..largest_factor).to_a
  potential_factor_array.each do |potential_factor|
    factor_array.each do |factor|
      if factor != potential_factor && factor % potential_factor == 0
        factor_array.delete(factor)
      end
    end
  end
  p factor_array.last
end

find_largest_prime(600851475143)
#=> 6857
